Add task_group::wait_for_task documentation (#2048)
oneapi-src/oneTBB: doc/main/reference/task_group_ext/task_completion_handle_cls.rst

.. _task_completion_handle_cls:

``task_completion_handle`` Class
================================

.. note::
    To enable this extension, define the ``TBB_PREVIEW_TASK_GROUP_EXTENSIONS`` macro with a value of ``1``.

.. contents::
    :local:
    :depth: 2

Description
***********

The ``task_completion_handle`` class represents a handle for the completion of the task

> created by ``task_group::defer``. Unlike ``task_handle``, which becomes empty once the task

Reviewer: ```suggestion
An instance of the ``task_completion_handle`` class represents a task for the purpose of
setting execution dependencies and tracking task completion.
Unlike ``task_handle``, which becomes empty once the task
```

is submitted for execution, a ``task_completion_handle`` keeps referencing a task

> regardless of whether it has been submitted, executing, or completed.

Reviewer: ```suggestion
regardless of whether it is submitted, executing, or completed.
```

.. code:: cpp

    tbb::task_group tg;

    tbb::task_handle th = tg.defer(task_body);
    // task is not submitted
    // th is non-empty and represents the task

    tbb::task_completion_handle tch = th;
    // task is not submitted
    // both th and tch are non-empty and represent the task

    tg.run(std::move(th));
    // task is submitted
    // th is empty
    // tch is non-empty and keeps representing the task

    tg.wait();
    // task is completed
    // tch is non-empty and represents the completed task

> This makes ``task_completion_handle`` useful for:

Reviewer: ```suggestion
The ``task_completion_handle`` class is used for:
```

* Establishing :ref:`dynamic task dependencies <dynamic_dependencies>`: it can serve as
  a predecessor in ``task_group::set_task_order`` to add successors even when the corresponding

>   task is submitted, executing, or completed.

Reviewer: ```suggestion
* Establishing :ref:`dynamic task dependencies <dynamic_dependencies>`: it can serve as
  a predecessor in ``task_group::set_task_order`` to add successors at any time, including
  after the corresponding task has been submitted and even completed.
```

* :ref:`Waiting for the completion of an individual task <wait_single_task>` using ``task_group::wait``,
  without waiting for all tasks in the group to finish.

A non-empty ``task_completion_handle`` can be obtained by constructing or assigning from a
non-empty ``task_handle`` or by copying another non-empty ``task_completion_handle``.
Multiple ``task_completion_handle`` objects may reference the same task simultaneously.
An empty ``task_completion_handle`` does not reference any task.

API
***

Header
------

.. code:: cpp

    #define TBB_PREVIEW_TASK_GROUP_EXTENSIONS 1
    #include <oneapi/tbb/task_group.h>

>     #include <oneapi/tbb/task_arena.h>

Reviewer: Unlike other cases where it is more or less clear which part of the synopsis is defined in which header, here two headers seem somewhat ambiguous: are both needed, or are both defining the class? I do not know how to deal best with that ambiguity; so far I prefer that only `task_group.h` is listed.

Reviewer 2: I think the same approach (listing only `task_group.h`) was used for `task_handle` when it was a part of reference.

Reviewer 2: Left only `task_group.h`
